src/logic/safe/store/actions/TransactionSender.ts

```diff
@@ -0,0 +1,259 @@
+    shortName: extractShortChainName(),
+    safeAddress,
+  })
+  const txRoute = generatePath(SAFE_ROUTES.TRANSACTIONS_SINGULAR, {
+    [SAFE_ADDRESS_SLUG]: prefixedSafeAddress,
+    [TRANSACTION_ID_SLUG]: txDetails.detailedExecutionInfo.safeTxHash,
+  })
+
+  history.push(txRoute)
+}
+
+type SenderArgs = {
+  props: CreateTransactionArgs | ProcessTransactionArgs
+  origin: string | null
+  dispatch: Dispatch
+  isExecuting: boolean
+  provider: ProviderState
+  safeVersion: string
+  txArgs: TxArgs
+  safeTxHash: string
+  safeInstance: GnosisSafe
+  navigateToDeeplink?: boolean
+  txId?: string
+}
+
+export type ConfirmEventHandler = (safeTxHash: string) => void
+export type ErrorEventHandler = () => void
+
+type TxHash = string | undefined
+
+export class TxSender {
+  origin: string | null
+  dispatch: Dispatch
+  isExecuting: boolean
+  provider: ProviderState
+  safeVersion: string
+  txArgs: TxArgs
+  safeTxHash: string
+  safeInstance: GnosisSafe
+  safeAddress: string
+  from: string
+  ethParameters: (CreateTransactionArgs | ProcessTransactionArgs)['ethParameters']
+  notifications: ReturnType<typeof createTxNotifications>
+
+  navigateToDeeplink?: boolean
+  txId?: string
+  confirmCallback?: ConfirmEventHandler
+  errorCallback?: ErrorEventHandler
+
+  constructor(
+    {
+      props,
+      origin,
+      dispatch,
+      isExecuting,
+      provider,
+      safeVersion,
+      txArgs,
+      safeTxHash,
+      safeInstance,
+      navigateToDeeplink,
+      txId,
+    }: SenderArgs,
+    confirmCallback?: ConfirmEventHandler,
+    errorCallback?: ErrorEventHandler,
+  ) {
+    this.origin = origin
+    this.dispatch = dispatch
+    this.isExecuting = isExecuting
+    this.provider = provider
+    this.safeVersion = safeVersion
+    this.txArgs = txArgs
+    this.safeTxHash = safeTxHash
+    this.safeInstance = safeInstance
+    this.safeAddress = props.safeAddress
+    this.from = provider.account
+    this.ethParameters = props.ethParameters
+    this.notifications = createTxNotifications(props.notifiedTransaction, origin, dispatch)
+
+    // Can be undefined
+    this.navigateToDeeplink = navigateToDeeplink
+    this.txId = txId
+    this.confirmCallback = confirmCallback
+    this.errorCallback = errorCallback
+  }
+
+  async submitTx(): Promise<TxHash> {
+    const { isExecuting, provider, safeVersion } = this
+
+    const canSignOffChain = checkIfOffChainSignatureIsPossible(isExecuting, provider.smartContractWallet, safeVersion)
+    if (!isExecuting && canSignOffChain) {
+      this.signOffChain()
+      return
+    }
+
+    const txHash = this.sendTx()
+    return txHash
+  }
+
+  private async signOffChain(): Promise<void> {
+    const { safeTxHash, txArgs, safeAddress, provider, safeVersion } = this
+
+    try {
+      const signature = await tryOffChainSigning(
+        safeTxHash,
+        { ...txArgs, safeAddress },
+        provider.hardwareWallet,
+        safeVersion,
+      )
+
+      this.onComplete(signature)
```

Comment:
I would move `onComplete` back to `submitTx`. The same for `onComplete` in the signing method.